x-pack/plugins/security/server/authentication/index.ts

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Boom from 'boom';
+import {
+  ClusterClient,
+  CoreSetup,
+  KibanaRequest,
+  LoggerFactory,
+  SessionStorageFactory,
+} from '../../../../../src/core/server';
+import { AuthenticatedUser } from '../../common/model';
+import { ConfigType } from '../config';
+import { getErrorStatusCode, wrapError } from '../errors';
+import { Authenticator, ProviderSession } from './authenticator';
+import { LegacyAPI } from '../plugin';
+
+export { canRedirectRequest } from './can_redirect_request';
+export { Authenticator, ProviderLoginAttempt } from './authenticator';
+export { AuthenticationResult } from './authentication_result';
+export { DeauthenticationResult } from './deauthentication_result';
+export { BasicCredentials } from './providers';
+
+export interface SetupAuthenticationParams {
+  core: CoreSetup;
+  clusterClient: PublicMethodsOf<ClusterClient>;
+  config: ConfigType;
+  loggers: LoggerFactory;
+  getLegacyAPI(): LegacyAPI;
+}
+
+export async function setupAuthentication({
+  core,
+  clusterClient,
+  config,
+  loggers,
+  getLegacyAPI,
+}: SetupAuthenticationParams) {
+  const authLogger = loggers.get('authentication');
+
+  const isSecurityFeatureDisabled = () => {
+    const xpackInfo = getLegacyAPI().xpackInfo;
+    return xpackInfo.isAvailable() && !xpackInfo.feature('security').isEnabled();
+  };
+
+  /**
+   * Retrieves currently authenticated user associated with the specified request.
+   * @param request
+   */
+  const getCurrentUser = async (request: KibanaRequest) => {
+    if (isSecurityFeatureDisabled()) {
+      return null;
+    }
+
+    return (await clusterClient
+      .asScoped(request)
+      .callAsCurrentUser('shield.authenticate')) as AuthenticatedUser;
+  };
+
+  const { sessionStorageFactory } = (await core.http.registerAuth<ProviderSession>(
+    async (request, t) => {
+      if (!authenticator) {
```

Comment:
The circular reference between the authenticationHandler and the authenticator is somewhat awkward, but I don't have any easy suggestions off-hand which would improve the situation.

Reply:
Yeah, I agree. I don't think we have a solution from the consumer side, but maybe platform can separate `sessionStorageFactory` from the `registerAuth` so that we can get sessionStorage factory, create authenticator and then register auth handler? @restrry what do you think?

Reply:
From a technical point of view, a session storage creation is not a requirement to define auth logic. My assumption that the current form can simplify usage failed. We can decouple them. I will update interfaces today.

Reply:
https://github.com/elastic/kibana/pull/40852